### PR TITLE
swarm: repeated --ens-api CLI flag (#14386)

### PR DESCRIPTION
**Pull request by janos**
This PR implement a CLI flag that can be repeated to allow multiple ENS
resolvers for different TLDs.

New flag --ens-endpoint can be repeated and accepts value with
format [tld:][contract-addr@]url.

MultiResolver is implemented to support resolutions based on their TLDs.
Each TLD can have multiple resolvers, and the resolution from the
first one in the sequence will be returned. If the TLD does not have its own resolver, a default resolver will be used.

ENSClientConfig is implemented to provide configuration parameters for ens resolvers on Swarm initialization.

Function detectEnsAddr has been moved from cmd/swarm/main.go to swarm/swarm.go.

### CODE REVIEW

**Comment by lmars** [comment]
Should this replace the `--ens-api` flag?

**Comment by janos** [comment]
Hi Lewis, since the flag --ens-endpoint can be repeated and can contain only the endpoint url/path, flags it can be simplified, that there is only one flag --ens-api which would function the same way as --ens-endpoint does now, just with the same default value, as the --ens-api does have now. One small problem is with --ens-addr, which can not be used with repeated --ens-api flag. I can "merge" --ens-endpoint functionality into --ens-api (default value) flag, and restrict --ens-addr to be allowed only if --ens-api is specified once.

Would such change make sense?

**Comment by lmars** [comment]
@janos so personally, I think the command line interface should be for "simple" use cases (e.g. I just want to point Swarm at mainnet / ropsten / rinkeby etc. to resolve all addresses), and we should promote using a config file for more complex use cases (@holisticode is currently working on improving the config file).

A config file can be self documenting with comments and support structures like maps and arrays in a natural way, whereas command line flags are quite cumbersome and error prone.

If we do expand the CLI args, I'd at least like to see only one way to configure ENS (either multiple `--ens-endpoint` flags or a modified `--ens-api` flag).

**Comment by janos** [comment]
@lmars the latest change addresses the last part of you comment. Flag --ens-endpoint is removed and --ens-api can be repeated.

File: cmd/swarm/main.go [hunk 1]
```diff
@@ -491,34 +454,78 @@ func registerBzzService(ctx *cli.Context, stack *node.Node) {
...
+		switch len(ensAPIs) {
+		case 0:
+			ensClientConfigs = append(ensClientConfigs, swarm.ENSClientConfig{
+				Endpoint:        node.DefaultIPCEndpoint("geth"),
+				ContractAddress: ensAddr,
+			})
+		case 1:
+			// Check if only one --ens-api is specified in order to use --ens-addr value
+			// to preserve the backward compatibility with single --ens-api flag.
+			// Multiple
```
Comment (lmars) [inline cmd/swarm/main.go]: This comment is incomplete.

File: cmd/swarm/main.go [hunk 2]
```diff
@@ -101,10 +97,9 @@ var (
 		Name:  "sync",
 		Usage: "Swarm Syncing enabled (default true)",
 	}
-	EnsAPIFlag = cli.StringFlag{
+	EnsAPIFlag = cli.StringSliceFlag{
 		Name:  "ens-api",
-		Usage: "URL of the Ethereum API provider to use for ENS record lookups",
-		Value: node.DefaultIPCEndpoint("geth"),
+		Usage: "ENS API endpoint for a TLD and with contract address, can be repeated, format [tld:][contract-addr@]url",
```
Comment (lmars) [inline cmd/swarm/main.go]: I guess we should deprecate `--ens-addr` then?

File: cmd/swarm/main.go [hunk 3]
```diff
@@ -491,34 +454,78 @@ func registerBzzService(ctx *cli.Context, stack *node.Node) {
...
+		case 0:
+			ensClientConfigs = append(ensClientConfigs, swarm.ENSClientConfig{
+				Endpoint:        node.DefaultIPCEndpoint("geth"),
+				ContractAddress: ensAddr,
+			})
+		case 1:
+			// Check if only one --ens-api is specified in order to use --ens-addr value
+			// to preserve the backward compatibility with single --ens-api flag.
+			// Multiple
+			c := parseFlagEnsAPI(ensAPIs[0])
```
Comment (lmars) [inline cmd/swarm/main.go]: We currently treat `--ens-api ''` as an indication to disable the use of ENS so we should check for that case here (but to be honest, we should have an explicit flag like `--no-ens` instead).

File: cmd/swarm/main.go [hunk 4]
```diff
@@ -491,34 +454,78 @@ func registerBzzService(ctx *cli.Context, stack *node.Node) {
...
 
-		return swarm.NewSwarm(ctx, swapClient, ensClient, bzzconfig, swapEnabled, syncEnabled, cors)
+		return swarm.NewSwarm(ctx, swapClient, ensClientConfigs, bzzconfig, swapEnabled, syncEnabled, cors)
 	}
 	if err := stack.Register(boot); err != nil {
 		utils.Fatalf("Failed to register the Swarm service: %v", err)
 	}
 }
 
+func parseFlagEnsAPI(s string) swarm.ENSClientConfig {
```
Comment (lmars) [inline cmd/swarm/main.go]: Add a comment to describe what this function is doing.

File: swarm/api/api.go [hunk 5]
```diff
@@ -40,6 +41,69 @@ type Resolver interface {
...
+type MultiResolverOption func(*MultiResolver)
+
+// MultiResolverOptionWithResolver adds a Resolver to a list of resolvers
+// for a specific TLD. If TLD is an empty string, the resolver will be added
+// to the list of default resolver, the ones that will be used for resolution
+// of addresses which do not have their TLD resolver specified.
+func MultiResolverOptionWithResolver(r Resolver, tld string) MultiResolverOption {
+	return func(m *MultiResolver) {
+		if _, ok := m.resolvers[tld]; !ok {
+			m.resolvers[tld] = []Resolver{}
```
Comment (lmars) [inline swarm/api/api.go]: There is no need for this assignment.

File: swarm/api/api.go [hunk 6]
```diff
@@ -40,6 +41,69 @@ type Resolver interface {
...
+		}
+		m.resolvers[tld] = append(m.resolvers[tld], r)
+	}
+}
+
+// NewMultiResolver creates a new instance of MultiResolver.
+func NewMultiResolver(opts ...MultiResolverOption) (m *MultiResolver) {
+	m = &MultiResolver{
+		resolvers: map[string][]Resolver{},
```
Comment (lmars) [inline swarm/api/api.go]: Use `make(map[string][]Resolver)` to initialise the map (that is the convention in the go-ethereum codebase).

File: swarm/api/api.go [hunk 7]
```diff
@@ -40,6 +41,69 @@ type Resolver interface {
...
+	}
+	return m
+}
+
+// Resolve resolves address by choosing a Resolver by TLD.
+// If there are more default Resolvers, or for a specific TLD,
+// the Hash from the the first one which does not return error
+// will be returned.
+func (m MultiResolver) Resolve(addr string) (h common.Hash, err error) {
+	rs, _ := m.resolvers[""]
```
Comment (lmars) [inline swarm/api/api.go]: No need to assign `_` here.

File: swarm/api/api.go [hunk 8]
```diff
@@ -40,6 +41,69 @@ type Resolver interface {
...
+	if i := strings.LastIndex(addr, "."); i >= 0 {
+		rstld, ok := m.resolvers[addr[i+1:]]
```
Comment (lmars) [inline swarm/api/api.go]: This would be more readable using [`path.Ext`](https://golang.org/pkg/path/#Ext).

**Review by lmars — APPROVED** [review]

**Comment by janos** [comment]
@holisticode could you review this pull request as it is related in some parts to this PR https://github.com/ethereum/go-ethereum/pull/15548?

**Comment by janos** [comment]
@holisticode this PR is updated with the changes that you did regarding config file feature. Could you review the changes?

There is one new filed EnsDisabled in swarm/api.Config. If that is ok, would it make sense to have cli flag --ens-disabled (as @lmars already suggested) and env ver SWARM_ENS_DISABLED?

One backward incompatible change is replacing Config field *EnsApi string* with *EnsAPIs []string*. Config file feature has been added recently, but if backward compatibility is needed, *EnsApi* field can be added and handled.

**Comment by gbalint** [comment]
For the note: team decision is that we disable ENS by default

File: swarm/swarm_test.go [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+			}
+			if addr != x.addr {
+				t.Errorf("expected ContractAddress %q, got %q", x.addr.String(), addr.String())
+			}
+			if tld != x.tld {
+				t.Errorf("expected TLD %q, got %q", x.tld, tld)
+			}
+		})
+	}
+}
```
Comment (holisticode) [inline swarm/swarm_test.go:119]: Can we add a test for invalid format via command line or env var params? E.g. `--ens-api ":<addr>"` or  `--ens-api "test-tld:" ` ? Can you think of other invalid ones? Are we also testing with multiple addresses per TLD? With no default resolver?

File: cmd/swarm/main.go [hunk 10]
```diff
@@ -110,16 +105,11 @@ var (
 		Usage:  "Swarm Syncing enabled (default true)",
 		EnvVar: SWARM_ENV_SYNC_ENABLE,
 	}
-	EnsAPIFlag = cli.StringFlag{
+	EnsAPIFlag = cli.StringSliceFlag{
 		Name:   "ens-api",
-		Usage:  "URL of the Ethereum API provider to use for ENS record lookups",
+		Usage:  "ENS API endpoint for a TLD and with contract address, can be repeated, format [tld:][contract-addr@]url",
```
Comment (holisticode) [inline cmd/swarm/main.go:110]: Have you tried to see if it works when the option is provided via config file?

Comment (holisticode) [reply cmd/swarm/main.go:110]: What is the behavior if only addresses with TLDs are provided, do we end up without a default one? Is it even possible to identify the default one if no default has been provided?

File: swarm/api/api.go [hunk 11]
```diff
@@ -40,6 +42,66 @@ type Resolver interface {
...
+// can be found for the address.
+var errNoResolver = errors.New("no resolver")
+
+// MultiResolver is used to resolve URL addresses based on their TLDs.
+// Each TLD can have multiple resolvers, and the resoluton from the
+// first one in the sequence will be returned.
+type MultiResolver struct {
+	resolvers map[string][]Resolver
+}
+
```
Comment (holisticode) [inline swarm/api/api.go:67]: why do we need multiple resolvers per TLD if only the first resolution is returned?

File: swarm/swarm.go [hunk 12]
```diff
@@ -76,7 +81,7 @@ func (self *Swarm) API() *SwarmAPI {
 
 // creates a new swarm service instance
 // implements node.Service
-func NewSwarm(ctx *node.ServiceContext, backend chequebook.Backend, ensClient *ethclient.Client, config *api.Config, swapEnabled, syncEnabled bool, cors string) (self *Swarm, err error) {
+func NewSwarm(ctx *node.ServiceContext, backend chequebook.Backend, config *api.Config) (self *Swarm, err error) {
```
Comment (holisticode) [inline swarm/swarm.go:84]: Good simplifying the func interface here, thanks

File: swarm/swarm.go [hunk 13]
```diff
@@ -133,18 +138,18 @@ func NewSwarm(ctx *node.ServiceContext, backend chequebook.Backend, ensClient *e
...
-		log.Warn("No ENS, please specify non-empty --ens-api to use domain name resolution")
-	} else {
-		self.dns, err = ens.NewENS(transactOpts, config.EnsRoot, ensClient)
-		if err != nil {
-			return nil, err
+	if len(config.EnsAPIs) > 0 {
+		opts := []api.MultiResolverOption{}
+		for _, c := range config.EnsAPIs {
+			tld, endpoint, addr := parseEnsAPIAddress(c)
+			r, err := newEnsClient(endpoint, addr, config)
```
Comment (holisticode) [inline swarm/swarm.go:145]: I wonder if we should test for invalid formats before we actually try to connect to ENS? What do you think?

File: cmd/swarm/main.go (hunk 10 above)
Comment (janos) [reply cmd/swarm/main.go:110]: The toml configuration file is loading the values this way:

EnsAPIs = [
        "endpoint1",
        "test:endpoint2",
]

The ENS API endpoint configuration is implemented to be most explicit. If no default endpoint is provided, no assumption are made. Could there be a case that no default resolver is required?

File: swarm/api/api.go (hunk 11 above)
Comment (janos) [reply swarm/api/api.go:67]: As per requirement in this comment https://github.com/ethereum/go-ethereum/issues/14386#issuecomment-334378961 for sequence of resolvers.

File: swarm/swarm.go (hunk 12 above)
Comment (janos) [reply swarm/swarm.go:84]: Thanks. :)

File: swarm/swarm.go (hunk 13 above)
Comment (janos) [reply swarm/swarm.go:145]: Function parseEnsAPIAddress does not return error as the format is very simple, permissive in the endpoint part and assumes any value for tld and contract address. I would put format validation much sooner during execution, just after the config load in cmd/swarm.buildConfig. Would that be ok?

File: swarm/swarm_test.go (hunk 9 above)
Comment (janos) [reply swarm/swarm_test.go:119]: 👍 Will add test cases for invalid formats.
Test cases for multiple resolvers are handled in swarm/api.TestMultiResolver with various combinations.

File: swarm/swarm.go (hunk 13 above)
Comment (holisticode) [reply swarm/swarm.go:145]: Yeah I think that would be a good idea.

File: swarm/swarm_test.go (hunk 9 above)
Comment (holisticode) [reply swarm/swarm_test.go:119]: great  thanks.

File: cmd/swarm/main.go (hunk 10 above)
Comment (holisticode) [reply cmd/swarm/main.go:110]: I am concerned about the other way around; let's say user provides `.eth:0x....@geth.ipc`, what happens to `.test` and other TLDs?

Comment (janos) [reply cmd/swarm/main.go:110]: My view is that in that case, other domains except under .eth TLD should not be resolved, as that is configured by the user. In this particular case, one may assume that provided endpoint which is ti only one can be used for other TLDs. But what is the case if endpoints for two or more TLDs are provided, but not a default endpoint? One option is not to have the default resolver, and the order one is to add all resolvers as default resolvers. Maybe the later behaviour is a better solution?

Comment (holisticode) [reply cmd/swarm/main.go:110]: I am not sure I understand, are you saying that if explicit TLD resolvers are provided but no default resolver, that all are tried as a default resolver? I think that would probably be the best option. Let's have a second opinion. @lmars ? @homotopycolimit ?

Comment (janos) [reply cmd/swarm/main.go:110]: Yes, that is right. I would also like to hear more opinions.

Comment (lmars) [reply cmd/swarm/main.go:110]: I think if the user doesn't provide a default (i.e. a value without the `tld:` part), then names which don't match an endpoint should not be resolved.

For example, if I explicitly start swarm with:

```
--ens-api .eth:<addr>@geth.ipc --ens-api .test:<addr>@testnet/geth.ipc
```

I would not expect say `.foo` requests to be resolved, rather an error be returned like `no ENS endpoint configured to resolve .foo names`.

Comment (janos) [reply cmd/swarm/main.go:110]: @lmars the functionality is now similar, just the error returned from MultiResolver.Resolve should include the information about the TLD, currently it is just an error (var errNoResolver) with "no resolver" message.

Comment (holisticode) [reply cmd/swarm/main.go:110]: ok, so @janos when you have changed the error message please ping, then I can mark this as approved from my side. Thanks.

File: swarm/swarm_test.go (hunk 9 above)
Comment (janos) [reply swarm/swarm_test.go:119]: Validations for missing tld, contract address or url is added to cmd/swarm.validateConfg.

File: cmd/swarm/main.go (hunk 10 above)
Comment (janos) [reply cmd/swarm/main.go:110]: There is a new error type NoResolverError for more verbose error message.

**Review by holisticode — APPROVED** [review]

**Comment by janos** [comment]
Thanks @holisticode.

@lmars are you still approving the PR, even with the latest changes?

File: cmd/swarm/config.go [hunk 14]
```diff
@@ -101,6 +102,8 @@ func buildConfig(ctx *cli.Context) (config *bzzapi.Config, err error) {
 	config = envVarsOverride(config)
 	//override settings provided by command line
 	config = cmdLineOverride(config, ctx)
+	//validate configuration parameters
+	err = validateConfig(config)
```
Comment (lmars) [inline cmd/swarm/config.go:109]: This shadows the previous assignment of `err` which is now not being checked (we should probably avoid named return arguments unless they are really needed)

File: cmd/swarm/config.go [hunk 15]
```diff
@@ -194,12 +197,16 @@ func cmdLineOverride(currentConfig *bzzapi.Config, ctx *cli.Context) *bzzapi.Con
...
+		// preserve backward compatibility to disable ENS with --ens-api=""
+		if len(ensAPIs) == 1 && ensAPIs[0] == "" {
+			ensAPIs = nil
+		}
+		currentConfig.EnsAPIs = ensAPIs
 	}
 
-	if ensaddr := ctx.GlobalString(EnsAddrFlag.Name); ensaddr != "" {
+	if ensaddr := ctx.GlobalString(DeprecatedEnsAddrFlag.Name); ensaddr != "" {
 		currentConfig.EnsRoot = common.HexToAddress(ensaddr)
```
Comment (lmars) [inline cmd/swarm/config.go:213]: We should print a deprecation notice here.

Comment (janos) [reply cmd/swarm/config.go:213]: The depreciation message is handled in checkDeprecated function on line 319.

File: cmd/swarm/config.go (hunk 14 above)
Comment (janos) [reply cmd/swarm/config.go:109]: Fixed by returning right after the previous error assignment if it is not nil.

File: cmd/swarm/config.go [hunk 16]
```diff
@@ -97,10 +98,15 @@ func buildConfig(ctx *cli.Context) (config *bzzapi.Config, err error) {
 	config = bzzapi.NewDefaultConfig()
 	//first load settings from config file (if provided)
 	config, err = configFileOverride(config, ctx)
+	if err != nil {
+		return err
```
Comment (lmars) [inline cmd/swarm/config.go]: Shouldn't this be `return nil, err`?

Comment (janos) [reply cmd/swarm/config.go]: Oh right, fixed. :( Thank you!

**Comment by zelig** [comment]
this looks good to me. ready for PR on main repo?

**Comment by janos** [comment]
@zelig PR created https://github.com/ethereum/go-ethereum/pull/15748.

**Comment by gbalint** [comment]
merged to `ethereum/master`

**Pull request closed without merging**